dir/src/lib.rs

```diff
@@ -8,56 +8,65 @@ use std::sync::{Arc, Weak};
 use wasi_common::dir::{ReaddirCursor, ReaddirEntity};
 use wasi_common::file::{Advice, FdFlags, FileType, Filestat, OFlags};
 use wasi_common::{Error, ErrorExt, SystemTimeSpec, WasiDir, WasiFile};
-use wasmtime_vfs_file::File;
-use wasmtime_vfs_ledger::{InodeId, Ledger};
-use wasmtime_vfs_memory::{Inode, Link, Node, Open, State};
+use wasmtime_vfs_ledger::{DeviceId, InodeId, Ledger};
+use wasmtime_vfs_memory::{Link, Node, Open, State};
 
-pub struct Directory(Link<BTreeMap<String, Arc<dyn Node>>>);
+type NodeConstructor = Arc<dyn Sync + Send + Fn(Arc<dyn Node>) -> Arc<dyn Node>>;
```

Comment:
```suggestion
type NodeConstructor = Arc<dyn (Fn(Arc<dyn Node>) -> Arc<dyn Node>) + Send + Sync>;
```

Looked strange at first to begin with Sync and Send...

Reply:
This is very subjective - I always start with `Sync` and `Send` in all the bounds I write.
Is there a guideline about this anywhere?

What does stdlib do?

Reply:
Some examples:
```
library/alloc/src/task.rs
89:impl<W: Wake + Send + Sync + 'static> From<Arc<W>> for Waker {
101:impl<W: Wake + Send + Sync + 'static> From<Arc<W>> for RawWaker {
116:fn raw_waker<W: Wake + Send + Sync + 'static>(waker: Arc<W>) -> RawWaker {
118:    unsafe fn clone_waker<W: Wake + Send + Sync + 'static>(waker: *const ()) -> RawWaker {
127:    unsafe fn wake<W: Wake + Send + Sync + 'static>(waker: *const ()) {
133:    unsafe fn wake_by_ref<W: Wake + Send + Sync + 'static>(waker: *const ()) {
139:    unsafe fn drop_waker<W: Wake + Send + Sync + 'static>(waker: *const ()) {

library/alloc/src/collections/btree/set/tests.rs
534:    fn set<T: Sync>(v: &BTreeSet<T>) -> impl Sync + '_ {
538:    fn iter<T: Sync>(v: &BTreeSet<T>) -> impl Sync + '_ {
542:    fn into_iter<T: Sync>(v: BTreeSet<T>) -> impl Sync {
546:    fn range<T: Sync + Ord>(v: &BTreeSet<T>) -> impl Sync + '_ {
550:    fn drain_filter<T: Sync + Ord>(v: &mut BTreeSet<T>) -> impl Sync + '_ {
554:    fn difference<T: Sync + Ord>(v: &BTreeSet<T>) -> impl Sync + '_ {
558:    fn intersection<T: Sync + Ord>(v: &BTreeSet<T>) -> impl Sync + '_ {
562:    fn symmetric_difference<T: Sync + Ord>(v: &BTreeSet<T>) -> impl Sync + '_ {
566:    fn union<T: Sync + Ord>(v: &BTreeSet<T>) -> impl Sync + '_ {
577:    fn iter<T: Send + Sync>(v: &BTreeSet<T>) -> impl Send + '_ {
585:    fn range<T: Send + Sync + Ord>(v: &BTreeSet<T>) -> impl Send + '_ {
593:    fn difference<T: Send + Sync + Ord>(v: &BTreeSet<T>) -> impl Send + '_ {
597:    fn intersection<T: Send + Sync + Ord>(v: &BTreeSet<T>) -> impl Send + '_ {
601:    fn symmetric_difference<T: Send + Sync + Ord>(v: &BTreeSet<T>) -> impl Send + '_ {
605:    fn union<T: Send + Sync + Ord>(v: &BTreeSet<T>) -> impl Send + '_ {
```

I'm happy to start with `Send`, rather than `Sync`, which seems to be the most common. I don't understand why you'd want to put `Fn` bound first, but I'll make the change to keep this moving forward. I don't think this really matters